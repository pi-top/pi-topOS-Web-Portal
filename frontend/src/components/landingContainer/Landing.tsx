import React, { useState, useEffect } from 'react';
import cx from "classnames";
import styles from './Landing.module.css';

import LandingListContainer from "../landingList/LandingListContainer";
import LandingDetailContainer from "../landingDetail/LandingDetailContainer";
import { LandingPageElement } from '../landing_app/App';

export type Props = {
  pages: LandingPageElement[];
};

export default ({ pages }: Props) => {
  const [ selectedElement, setSelectedElement ] = useState<LandingPageElement | undefined >()

  useEffect(() => {
    pages && pages.length > 0 && setSelectedElement(pages[0]);
  }, [pages])

  return (
    <div className={cx(styles.container)}>
        { pages && selectedElement !== undefined && (
          <>
          <LandingListContainer
            pages={pages}
            activeElement={selectedElement}
            onClick={(element: LandingPageElement) => setSelectedElement(element)}
            className={styles.landingList}
          />

          <div className={styles.landingPage}>
            <LandingDetailContainer page={selectedElement}/>
          </div>

          </>
        )}
    </div>
  );
};
