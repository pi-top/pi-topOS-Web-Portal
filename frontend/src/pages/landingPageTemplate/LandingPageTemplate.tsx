import React, { useEffect, useState, ReactElement } from "react";

import Layout from "../../components/layout/Layout";
import Spinner from "../../components/atoms/spinner/Spinner";

import styles from "./LandingPageTemplate.module.css";
import { runningOnWebRenderer } from "../../helpers/utils";
import closePtOsLandingWindow from "../../services/closePtOsLandingWindow";

export type UrlData = {
  defaultUrl: string;
  urlService?: () => Promise<{ [s: string]: string }>;
  onWebRenderer: () => Promise<void>;
};

export type Props = {
  title: string;
  message: React.ReactNode;
  prompt: ReactElement;
  image: string;
  urlInfo: UrlData;
  buttonLabel?: string;
};

export default ({ ...props }: Props) => {
  const [isOpeningUrl, setIsOpeningUrl] = useState(false);
  const [url, setUrl] = useState<string>(props.urlInfo.defaultUrl);
  const [, setError] = useState(false);

  useEffect(() => {
    if (props.urlInfo.urlService) {
      props.urlInfo.urlService()
        .then((data: any) => setUrl(data.url))
        .catch(() => setUrl(props.urlInfo.defaultUrl))
    }
  }, [props]);

  return (
    <Layout
      banner={{
        src: props.image,
        alt: "banner"
      }}
      prompt={props.prompt}
      nextButton={{
        onClick: () => {
          if (runningOnWebRenderer()) {
            setIsOpeningUrl(true);
            props.urlInfo.onWebRenderer()
              .then(() => {
                setIsOpeningUrl(true);
                window.setTimeout(() => {
                  setIsOpeningUrl(false);
                  closePtOsLandingWindow();
                }, 10000);
              })
              .catch(() => setError(true))
            ;
          } else {
            window.open(url)
          }
        },
        label: props.buttonLabel? props.buttonLabel : "Let's Go",
        disabled: isOpeningUrl,
      }}

      className={styles.root}
      showBack={false}
      showSkip={false}
      showHeader={false}
    >
      <div className={styles.message}>
        {props.message}
        <br />
        {isOpeningUrl && <Spinner size={45} />}
      </div>
    </Layout>
  );
};
