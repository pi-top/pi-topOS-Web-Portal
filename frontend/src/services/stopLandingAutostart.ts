 import axios from "axios";

import apiBaseUrl from "./apiBaseUrl";

export default async function stopLandingAutostart() {
  await axios.post(
    `${apiBaseUrl}/disable-landing`,
    {}
  );
}
