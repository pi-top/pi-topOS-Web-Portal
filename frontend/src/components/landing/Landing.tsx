import React, { useEffect, useState } from "react";
import cx from "classnames";
import styles from "./Landing.module.css";

import Button from "../atoms/button/Button";

import stopLandingAutostart from "../../services/stopLandingAutostart";

export type LandingPageElement = {
  title: string;
  id: string;
  detail: JSX.Element;
};

export type Props = {
  pages: LandingPageElement[];
};

export default ({ pages }: Props) => {
  const [selectedElement, setSelectedElement] = useState(pages[0]);

  useEffect(() => {
    stopLandingAutostart().catch(() => null);
  }, []);

  return (
    <div className={cx(styles.container)}>
      <div className={styles.landingList}>
        {pages.map((element) => (
          <div key={element.id} className={styles.elementDiv}>
            <Button
              unstyled
              className={cx(
                styles.element,
                selectedElement.id === element.id
                  ? styles.selectedElement
                  : ""
              )}
              onClick={() => setSelectedElement(element)}
            >
              <span className={styles.elementText}>{element.title}</span>
            </Button>
          </div>
        ))}
      </div>

      <div className={styles.landingPage}>
        <div className={styles.detailContainer}>
          {selectedElement.detail}
        </div>
      </div>
    </div>
  );
};
