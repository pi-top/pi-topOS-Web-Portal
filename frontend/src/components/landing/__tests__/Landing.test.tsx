import React from "react";
import { fireEvent, getAllByText, render, wait, waitForElement } from "@testing-library/react";

import Landing, { Props } from "../Landing";
import LandingPageTemplate from "../../../pages/landingPageTemplate/LandingPageTemplate";
import stopLandingAutostart from "../../../services/stopLandingAutostart";

jest.mock("../../../services/stopLandingAutostart");

const stopLandingAutostartMock = stopLandingAutostart as jest.Mock;


const firstPageContent = {
  title: "This is the title of the first page",
  id: "kb",
  templateTitle: "This is the title for the first page",
  urlInfo: {
    defaultUrl: "https://www.pi-top.com",
    onWebRenderer: jest.fn(),
  },
  message: "A message for the first",
  prompt: <>This is a <span className="green">prompt</span></>,
  image: "Any image"
}

const secondPageContent = {
  title: "This is another title",
  id: "sdk",
  templateTitle: "This is title for the second page",
  urlInfo: {
    defaultUrl: "https://further.pi-top.com",
    onWebRenderer: jest.fn(),
  },
  message: "This is the message for the second page",
  prompt: <>This is the prompt for the <span className="green">second</span> page</>,
  image: "Second image"
}


const landingPages = [
  {
    title: firstPageContent.title,
    id: firstPageContent.id,
    detail: (
      <LandingPageTemplate
        key={firstPageContent.id}
        title={firstPageContent.templateTitle}
        urlInfo={firstPageContent.urlInfo}
        message={firstPageContent.message}
        prompt={firstPageContent.prompt}
        image={firstPageContent.image}
      />
    )
  },
  {
    title: secondPageContent.title,
    id: secondPageContent.id,
    detail: (
      <LandingPageTemplate
        key={secondPageContent.id}
        title={secondPageContent.templateTitle}
        urlInfo={secondPageContent.urlInfo}
        message={secondPageContent.message}
        prompt={secondPageContent.prompt}
        image={secondPageContent.image}
      />
    )
  },
]

describe("Landing", () => {
  let layout: HTMLElement;
  let defaultProps: Props;
  let queryByAltText: any;
  let queryByText: any;
  let queryByLabelText: any;
  let getByText: any;
  let rerender: any;
  beforeEach(() => {
    stopLandingAutostartMock.mockResolvedValue("OK");

    defaultProps = {
      pages: landingPages,
    };

    ({
      container: layout,
      queryByAltText,
      queryByText,
      queryByLabelText,
      getByText,
      rerender,
    } = render(<Landing {...defaultProps} />));
  });

  it("renders the list of pages", () => {
    expect(layout.querySelector(".landingList")).toBeInTheDocument();
  });

  it("renders all the provided pages in the list of pages", () => {
    landingPages.forEach((page) => {
      expect(queryByText(page.title)).toBeInTheDocument()
    });
  });

  it("first element of the list is set as active in the list of pages", () => {
    expect(layout.querySelector(".selectedElement")).toHaveTextContent(firstPageContent.title);
  });

  it("only one element of the list is set as active", () => {
    expect(layout.querySelectorAll(".selectedElement").length).toBe(1)
  });

  it("renders the active page image", () => {
    expect(queryByAltText("banner")).toMatchSnapshot();
  });

  it("renders the active page message", () => {
    expect(getByText(firstPageContent.message)).toBeInTheDocument();
  });

  it("renders the active page prompt", () => {
    expect(layout.querySelector(".prompt")).toMatchSnapshot();
  });

  it("stops landing from autostarting on boot", () => {
    expect(stopLandingAutostartMock).toHaveBeenCalled();
  });

  describe("when clicking another page from the list", () => {
    beforeEach(async () => {
      fireEvent.click(getByText(secondPageContent.title))
      wait();
    });

    it("clicked page is set as active in the list of pages", () => {
      expect(layout.querySelector(".selectedElement")).toHaveTextContent(secondPageContent.title);
    });

    it("only one element of the list is set as active", () => {
      expect(layout.querySelectorAll(".selectedElement").length).toBe(1)
    });

    it("renders the selected page image", () => {
      expect(queryByAltText("banner")).toMatchSnapshot();
    });

    it("renders the selected page message", () => {
      expect(getByText(secondPageContent.message)).toBeInTheDocument();
    });

    it("renders the selected page prompt", () => {
      expect(layout.querySelector(".prompt")).toMatchSnapshot();
    });
  });

});
