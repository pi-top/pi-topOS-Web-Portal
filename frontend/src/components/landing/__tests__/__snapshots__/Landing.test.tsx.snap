// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing renders the active page image 1`] = `
<img
  alt="banner"
  class="image"
  draggable="false"
  src="Any image"
  style="max-height: 100%; max-width: 100%;"
/>
`;

exports[`Landing renders the active page prompt 1`] = `
<h1
  class="prompt"
>
  This is a 
  <span
    class="green"
  >
    prompt
  </span>
</h1>
`;

exports[`Landing when clicking another page from the list renders the selected page image 1`] = `
<img
  alt="banner"
  class="image"
  draggable="false"
  src="Second image"
  style="max-height: 100%; max-width: 100%;"
/>
`;

exports[`Landing when clicking another page from the list renders the selected page prompt 1`] = `
<h1
  class="prompt"
>
  This is the prompt for the 
  <span
    class="green"
  >
    second
  </span>
   page
</h1>
`;
